flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java

```diff
@@ -808,16 +850,31 @@ private void fulfillPendingSlotRequestWithPendingTaskManagerSlot(PendingSlotRequ
 		return Optional.empty();
 	}
 
-	private boolean isFulfillableByRegisteredSlots(ResourceProfile resourceProfile) {
+	private boolean isFulfillableByRegisteredOrPendingSlots(ResourceProfile resourceProfile) {
 		for (TaskManagerSlot slot : slots.values()) {
 			if (slot.getResourceProfile().isMatching(resourceProfile)) {
 				return true;
 			}
 		}
+
+		for (PendingTaskManagerSlot slot : pendingSlots.values()) {
+			if (slot.getResourceProfile().isMatching(resourceProfile)) {
+				return true;
+			}
+		}
+
 		return false;
 	}
 
 	private Optional<PendingTaskManagerSlot> allocateResource(ResourceProfile requestedSlotResourceProfile) {
+		final int numRegisteredSlots =  getNumberRegisteredSlots();
+		final int numPendingSlots = getNumberPendingTaskManagerSlots();
+		if (numPendingSlots + numRegisteredSlots + numSlotsPerWorker > maxSlotNum) {
+			LOG.warn("Could not allocate {} more slots. The number of registered and pending slots is {}, while the maximum is {}.",
+				numSlotsPerWorker, numPendingSlots + numRegisteredSlots, maxSlotNum);
+			return Optional.empty();
+		}
```

Comment:
nit: potentially this can be deduplicated with `isMaxSlotNumExceededAfterRegistration` if we make it something like:
`isMaxSlotNumExceededAfterRegistration(int maxNewSlotsNumber, Supplier<Integer> newSlotsNumberExact)`
but maybe too complicated.

Reply:
I think it probably makes sense to add
```
private boolean isMaxSlotNumExceededAfterAdding(int numNewSlot);
```
WDYT?

Reply:
do you mean to add just `isMaxSlotNumExceededAfterAdding` without deduplication with `isMaxSlotNumExceededAfterRegistration`? because `isMaxSlotNumExceededAfterRegistration` has 2 checks against 2 different `numNewSlot`.